src/shrink.rs

```diff
@@ -25,3 +27,65 @@ pub fn shrink(m: &mut Module) {
     }
     passes::gc::run(m);
 }
+
+pub fn optimize(m: &mut Module, keep_name_section: bool, level: &str) -> anyhow::Result<()> {
+    // recursively optimize embedded modules in Motoko actor classes
+    if is_motoko_canister(m) {
+        let data = get_motoko_wasm_data_sections(m);
+        for (id, mut module) in data.into_iter() {
+            optimize(&mut module, keep_name_section, level)?;
+            let blob = encode_module_as_data_section(module);
+            m.data.get_mut(id).value = blob;
+        }
+    }
+
+    // write module to temp file
+    let temp_file = NamedTempFile::new()?;
+    m.emit_wasm_file(temp_file.path())?;
+
+    // pull out a copy of the custom sections to preserve
+    let mut metadata_sections: Vec<(Kind, &str, Vec<u8>)> = Vec::new();
+    list_metadata(m).iter().for_each(|full_name| {
```

Comment:
Let's inline the code for `list_metadata` here. `collection().iter()` and `get_metadata` is all redundant.